src/i2c.c

```diff
@@ -0,0 +1,277 @@
+
+    // Send START condition.
+    i2c_send_start(I2C1);
+
+    // Waiting for START to send and switch to master mode.
+    while (!((I2C_SR1(I2C1) & I2C_SR1_SB)
+        && (I2C_SR2(I2C1) & (I2C_SR2_MSL | I2C_SR2_BUSY))));
+
+    // Say to what address we want to talk to.
+    i2c_send_7bit_address(I2C1, addr, I2C_WRITE);
+
+    // Waiting for address to transfer.
+    while (!(I2C_SR1(I2C1) & (I2C_SR1_ADDR | I2C_SR1_AF)));
+    I2C_FAIL_ON_NACK();
+
+    // Cleaning ADDR condition sequence.
+    reg32 = I2C_SR2(I2C1);
+}
+
+void i2c_stop_message(void) {
+    I2C_EXIT_ON_FAILED();
+
+    // Wait for the data register to be empty or a NACK to be generated.
+    while (!(I2C_SR1(I2C1) & (I2C_SR1_TxE | I2C_SR1_AF)));
+    I2C_FAIL_ON_NACK();  /// TODO Is a NACK expected here?
+
+    // Send STOP condition.
+    i2c_send_stop(I2C1);
+}
+
+void i2c_send_byte(char c) {
+    I2C_EXIT_ON_FAILED();
+
+    i2c_send_data(I2C1, c);
+    // Wait for byte to complete transferring
+    while (!(I2C_SR1(I2C1) & (I2C_SR1_BTF | I2C_SR1_AF)));
+    I2C_FAIL_ON_NACK();
+}
+
+bool i2c_recv_bytes(uint8_t addr, uint8_t* buf, uint8_t len) {
+    uint32_t reg32 __attribute__((unused));
+    I2C_EXIT_ON_FAILED(false);
+
+    if (len == 0) {
+        return false;
+    }
+
+    // Send START condition.
+    i2c_send_start(I2C1);
+
+    // Waiting for START to send and switch to master mode.
+    while (!((I2C_SR1(I2C1) & I2C_SR1_SB)
+        && (I2C_SR2(I2C1) & (I2C_SR2_MSL | I2C_SR2_BUSY))));
+
+    // Say to what address we want to read from.
+    i2c_send_7bit_address(I2C1, addr, I2C_READ);
+
+    if (len == 1) {
+        // clear the ACK bit.
+        i2c_disable_ack(I2C1);
+
+        // Waiting for address to transfer.
+        while (!(I2C_SR1(I2C1) & (I2C_SR1_ADDR | I2C_SR1_AF)));
+        I2C_FAIL_ON_NACK(false);
+
+        // Clear ADDR
+        reg32 = I2C_SR2(I2C1);
+
+        // Program the STOP bit.
+
+        // Read the data after the RxNE flag is set.
+        while (!(I2C_SR1(I2C1) & (I2C_SR1_RxNE | I2C_SR1_AF)));
+        I2C_FAIL_ON_NACK(false);
+
+        buf[0] = i2c_get_data(I2C1);
+    } else if (len == 2) {
+        // Set POS and ACK
+        i2c_enable_ack(I2C1);
+        i2c_nack_next(I2C1);
+
+        // Waiting for address to transfer.
+        while (!(I2C_SR1(I2C1) & (I2C_SR1_ADDR | I2C_SR1_AF)));
+        I2C_FAIL_ON_NACK(false);
+
+        // Clear ADDR
+        reg32 = I2C_SR2(I2C1);
+
+        // Clear ACK
+        i2c_disable_ack(I2C1);
+
+        // Wait for BTF to be set
+        while (!(I2C_SR1(I2C1) & (I2C_SR1_BTF | I2C_SR1_AF)));
+        I2C_FAIL_ON_NACK(false);
+
+        // Program STOP
+        i2c_send_stop(I2C1);
+
+        // Read DR twice
+        buf[0] = i2c_get_data(I2C1);
+        buf[1] = i2c_get_data(I2C1);
+
+        // Wait for transaction to complete
+        while (I2C_SR2(I2C1) & I2C_SR2_BUSY);
+        // Reset NACK control
+        i2c_nack_current(I2C1);
+    } else {  /// TODO this locks up
```

Comment:
Sounds bad - is this issue tracked somewhere?

Reply:
Tracked in #14